### ISSUE TITLE
Azure Function Custom Trigger Base on Azure Redis

### ISSUE THREAD

**Issue by loges16** [Azure/Azure-Functions#658]
Hi 
I need to trigger an function whenever an update/Insert happens in Azure redis cache
is it Possible? to create a custom trigger for redis cache

**Comment by jeffhollan** [comment]
Nothing there today that will do this.  The best way to do a custom trigger currently is have something that can poll or subscribe to redis cache and publish an event to either a queue/topic or event grid when an event occurs.  Then you can start building functions that listen to that event grid.  One option is a timer based function that wakes up every few seconds and checks Redis, but not sure if the latency of "polling" would be too much.  Other option would be some long-running container or VM or process that is subscribed to redis and can publish events.

May be worth adding an ask on our Uservoice for Redis Cache trigger - though I'm not certain how simple that would be to implement given how "close to the bits" Redis SDKs tend to work.  Though looks like you beat me to it [here](https://feedback.azure.com/forums/355860-azure-functions/suggestions/32980363-functions-should-support-custom-trigger-events) 👍

**Comment by ChristianWeyer** [comment]
@jeffhollan Wouldn't it be possible to take a similar approach like here: https://medium.com/@yuka1984/how-to-make-custom-trigger-with-azure-functions-3bd85551ac5e ?

Thanks.

**Comment by forki** [comment]
would love to see redis subscription trigger. It would be super helpful and it would works nicely with your Azure Redis service.

**Comment by jeffhollan** [comment]
Agree - but not sure how we could pull it off looking at Redis APIs.  Doesn't seem to be a good way to get change notifications on a redis server. Closest I can see is [the keyspace notification](https://redis.io/topics/notifications) but that doesn't provide reliable notifications and requires constant listening of events (and server config to send events) which our other triggers don't do in same way.  I think best thing to do would be keep voting for the uservoice item linked above, it's a valid ask, but I don't see an 'easy' way to implement this with current Redis version

**Comment by forki** [comment]
https://redis.io/commands/subscribe is the command that is interesting here. Subscribing to channels.

**Comment by jeffhollan** [comment]
Yes but appears to suffer from same limitations in that you only get messages if are actively subscribed and constantly polling. Also that would be limited to only triggering on PUBLISH commands and initial ask was for trigger on insert/publish of keys

**Comment by forki** [comment]
Yes the runtime would need to do the subscriptions that's the whole point and the only way how this would make sense in a consumption plan. But it would extremely useful. Especially given that we buy azure functions and redis hosting from Microsoft

**Comment by patriziobruno** [comment]
@jeffhollan, Redis' aren't key-space-notifications similar in concept to servicebus?

**Comment by jeffhollan** [comment]
I think a big distinction would be what I just read in the docs for this:

> Because Redis Pub/Sub is fire and forget currently there is no way to use this feature if your application demands reliable notification of events, that is, if your Pub/Sub client disconnects, and reconnects later, all the events delivered during the time the client was disconnected are lost

That’s a bit tricky for functions where they are offline by design for periods at a time and expected to wake up and retrieve a message after getting a notification. 

But I suspect there may be something here - just don’t know enough about redis if there’s some “task queue” we could pull from after getting a notification

**Comment by forki** [comment]
The function is offline - the function runtime is not

Jeff Hollan <notifications@github.com> schrieb am Sa., 27. Apr. 2019, 20:42:

**Comment by jeffhollan** [comment]
> The function is offline - the function runtime is not

The function and the runtime both go offline.  We have an additional component called the scale controller which can monitor your event source and activate your function if an event is present, but it's not a persistent connection.  It checks for events briefly on a source, and then disconnects and checks again after some interval.

**Comment by dannyhulmston** [comment]
Is there any plans by the Azure Functions team to provide an Azure Cache for Redis trigger? I'm currently working a contract where the main source of data is Cosmos DB and we are using the Cosmos DB trigger to replicate the data to the cache but the dev team want to update the cache and have that push an update back to Cosmos DB.

**Comment by gorillapower** [comment]
Checking back into this thread to check if there is azure function trigger support for redis? I believe its not possible to create a consumption plan based custom trigger at this time, unless new releases have made this possible?

**Comment by MSFTeegarden** [comment]
@gorillapower , @dh-technology-co-uk , @forki 

Update: there is now [Azure Function Trigger support for Redis](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-cache?tabs=isolated-process&pivots=programming-language-csharp)

Pubsub channels (including keyspace and keyevent notifications) are supported, as are Lists and Streams. [Here's the actual repo for the functions extension](https://github.com/Azure/azure-functions-redis-extension). 

And [here is a basic getting started tutorial](https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-tutorial-functions-getting-started), plus a more advanced [write-behind tutorial using Azure SQL DB](https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-tutorial-write-behind).

**Comment by willvelida** [comment]
> @gorillapower , @dh-technology-co-uk , @forki
> 
> Update: there is now [Azure Function Trigger support for Redis](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-cache?tabs=isolated-process&pivots=programming-language-csharp)
> 
> Pubsub channels (including keyspace and keyevent notifications) are supported, as are Lists and Streams. [Here's the actual repo for the functions extension](https://github.com/Azure/azure-functions-redis-extension).
> 
> And [here is a basic getting started tutorial](https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-tutorial-functions-getting-started), plus a more advanced [write-behind tutorial using Azure SQL DB](https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-tutorial-write-behind).

Awesome work @MSFTeegarden! Don't suppose I can ask when Consumption plan functions will be supported? 👀

**Comment by MSFTeegarden** [comment]
Hi @willvelida! Hope you're doing well. 😊 We plan on supporting Consumption functions in the march timeframe. Note that consumption functions won't help very much with the pubsub trigger since Redis pubsub is fire and forget. 

Bindings are also supported now, by the way. There is documentation in the extension repo, although we haven't added to our public Azure docs yet.
